### PR TITLE
Frontend/full height course cards

### LINKED ISSUE #377: (idea) Add option to "focus" on CourseCard, making it take 100% of height
I think there should be an option to make a given CourseCard take up the entire height of the CourseSelectColumn (first column), as it might be useful when you're doing the Section customization level. We can discuss if this is actually a feature we want or not (hence the `(idea)` in title) - I don't think it's something we need before we launch

### PR DESCRIPTION
**Pull request by ryan-conn**
## Description

Moves expand/collapse for course cards to redux, only one can be expanded at a time, and the expanded card will fill as much height as possible (with a minimum height roughly equal to what it was previously). No tests right now because I wanted some input on if we like how this works, specific behavior, etc.

## Rationale
The only thing in this branch that isn't how I imagined it is the extra height when the currently selected card is in section view and only has a couple sections. I couldn't think of any way to keep the fit-to-height behavior as well as having a minimum height without making this compromise (despite spending 4 hours in CSS hell), but I still think that in the vast majority of use cases this is a lot nicer.

## How to test
Make some cards, the expanded one will fit to your screen size (unless you're Adel and have a screen made for ants, in which case I've made it so this works basically the same way as before)

## Screenshots
I think that 7 course cards is about as high as most users will go, so here are some example sizes:

| Old appearance (always 300px) | 1080, 5 cards (417px) | 1080p, 7 cards (340px) |
|--|--|--|
| image | image | image |

| 1440p, 7 cards (697px) | 720p (314px, any more than 3 cards makes it scroll) |
|--|--|
| image | image |

Finally, here's the one case I don't really like the appearance of. If you're on the section select and there are only 1 or 2 sections, there will be some extra empty space that I couldn't find any way to avoid even using refs:
image
It looks normal on basic select though since I check if you're on section select with a course selected before giving a minimum height:
image

## Related tasks

Closes #377.

### CODE REVIEW

**Comment by ryan-conn** [comment]
Should be ready for review now that the edge case with low height section select has been fixed and I added tests.

**Comment by firejake308** [comment]
I want this merged because I am so tired of manually collapsing course cards

**Comment by gannonprudhomme** [comment]
> I want this merged because I am so tired of manually collapsing course cards

Then review it 😉

**Comment by ryan-conn** [comment]
Changes I just pushed make it so if you delete the expanded card, it will expand one next to it (below if possible, or above if none are below it)

**Comment by gannonprudhomme** [comment]
Bug I ran into:
full-course-cards-bug

**Comment by ryan-conn** [comment]
> Bug I ran into:

Fixed, this was due to destructuring the existing state instead of the new card attributes when `addCourseCard` is called with an existing card.

I also noticed making this change that it will *always* expand the new card (even when loading your cards, when you don't want to expand them), so I fixed that as well

File: autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectColumn.tsx [hunk 1]
```diff
@@ -19,10 +20,20 @@ const CourseSelectColumn: React.FC = () => {
...
-
+  const term = useSelector<RootState, string>((state) => state.term);
   const dispatch = useDispatch();
 
-  const term = useSelector<RootState, string>((state) => state.term);
+  const expandedRowRef = React.useRef<HTMLDivElement>(null);
+  React.useLayoutEffect(() => {
+    if (expandedRowRef.current) {
+      const expandedRowHeight = expandedRowRef.current.children[0].clientHeight;
+      if (expandedRowHeight < 500 - 8) {
```
Comment (firejake308) [inline autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectColumn.tsx:30]: We should probably mention that 500 needs to be the minimum height from the `.expanded-row` rule, and that the 8 comes from the `padding-top` on `.row`

File: autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectColumn.css [hunk 2]
```diff
@@ -13,25 +14,35 @@
...
+}
+
+.expanded-row {
+    overflow-y: hidden;
+    min-height: 500px;
 }
 
-.row:nth-last-of-type(2) {
-    padding-bottom: 0px;
+.expanded-row-overflow {
```
Comment (firejake308) [inline autoscheduler/frontend/src/components/SchedulingPage/CourseSelectColumn/CourseSelectColumn.css:44]: I would rename this to `expanded-row-small` since it's applied to small/short cards that fit within the 500px limit

**Review by firejake308 — CHANGES_REQUESTED** [review]
Tried to break the code in browser by mindlessly adding, deleting, expanding, and collapsing different course cards, no bugs found. Love `autoFocus` btw. 2 suggestions for comments/code style and this should be good to go. We should probably have Gannon test it too, though, since he's better at teasing out bugs than I am usually.

**Review by gannonprudhomme — COMMENTED** [review]
minor things I wrote a bit ago, will re-review later tn

**Comment by ryan-conn** [comment]
Comments should be addressed, updated with master

File: autoscheduler/frontend/src/redux/reducers/courseCards.ts [hunk 3]
```diff
@@ -43,34 +43,97 @@ const initialCourseCardArray: CourseCardArray = {
...
+        return getStateAfterExpanding(state, action.index, action.courseCard);
+      }
+
+      // otherwise just update this card
       return {
         ...state,
         [action.index]: { ...state[action.index], ...action.courseCard },
         numCardsCreated: Math.max(state.numCardsCreated, action.index + 1),
       };
+    }
```
Comment (gannonprudhomme) [inline autoscheduler/frontend/src/redux/reducers/courseCards.ts:136]: Don't think you need the braces around this case block, as you don't declare anything

**Review by gannonprudhomme — CHANGES_REQUESTED** [review]
Few minor things, otherwise this looks good to me; also couldn't find any bug when I tried messing with it
